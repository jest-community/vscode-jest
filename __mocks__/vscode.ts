const languages = {
  createDiagnosticCollection: jest.fn(),
}

const StatusBarAlignment = {}

const window = {
  createStatusBarItem: jest.fn(() => ({
    show: jest.fn(),
  })),
  showErrorMessage: jest.fn(),
  createTextEditorDecorationType: jest.fn(),
}

const workspace = {
  getConfiguration: jest.fn(),
  workspaceFolders: jest.fn(),
}

const OverviewRulerLane = {
  Left: null,
}

const Uri = { file: f => f }
const Range = jest.fn()
const Diagnostic = jest.fn()
const DiagnosticSeverity = { Error: 0, Warning: 1, Information: 2, Hint: 3 }

const debug = {
  onDidTerminateDebugSession: jest.fn(),
  startDebugging: jest.fn(async (_folder: any, nameOrConfig: any) => {
    if (typeof nameOrConfig === 'string') {
      throw null
    }
  }),
}

export {
  languages,
  StatusBarAlignment,
  window,
  workspace,
  OverviewRulerLane,
  Uri,
  Range,
  Diagnostic,
  DiagnosticSeverity,
  debug,
}
