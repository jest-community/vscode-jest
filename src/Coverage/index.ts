export * from './CoverageMapProvider';
export * from './CoverageOverlay';
export * from './CoverageCodeLensProvider';
