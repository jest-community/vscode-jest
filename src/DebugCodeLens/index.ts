export { DebugCodeLensProvider } from './DebugCodeLensProvider'
export { TestState } from './TestState'
