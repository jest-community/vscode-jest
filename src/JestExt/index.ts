export * from './core';
export * from './types';
export * from './auto-run';
