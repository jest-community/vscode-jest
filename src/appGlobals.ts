export const extensionName = 'io.orta.jest'
export const extensionId = 'orta.vscode-jest'
