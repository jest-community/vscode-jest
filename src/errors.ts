export interface ExtErrorDef {
  code: number;
  type: 'error' | 'warn';
  desc: string;
  helpLink: string;
}

const BASE_URL = 'https://github.com/jest-community/vscode-jest/blob/master/README.md';
export const GENERIC_ERROR: ExtErrorDef = {
  code: 1,
  type: 'error',
  desc: 'jest test run failed',
  helpLink: `${BASE_URL}#troubleshooting`,
};
export const CMD_NOT_FOUND: ExtErrorDef = {
  code: 2,
  type: 'error',
  desc: 'jest process failed to start, most likely due to env or project configuration issues',
  helpLink: `${BASE_URL}#jest-failed-to-run`,
};
export const LONG_RUNNING_TESTS: ExtErrorDef = {
  code: 3,
  type: 'warn',
  desc: 'jest test run exceed the configured threshold ("jest.monitorLongRun") ',
  helpLink: `${BASE_URL}#what-to-do-with-long-running-tests-warning`,
};

export const getExitErrorDef = (exitCode?: number): ExtErrorDef | undefined => {
  if (exitCode === 127) {
    return CMD_NOT_FOUND;
  }
};
