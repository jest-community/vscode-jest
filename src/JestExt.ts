import * as vscode from 'vscode'
import * as path from 'path'
import { Settings, ProjectWorkspace, JestTotalResults } from 'jest-editor-support'
import { matcher } from 'micromatch'

import * as decorations from './decorations'
import { IPluginSettings } from './IPluginSettings'
import * as status from './statusBar'
import {
  TestReconciliationState,
  TestResultProvider,
  TestResult,
  resultsWithLowerCaseWindowsDriveLetters,
} from './TestResults'
import { pathToJestPackageJSON, pathToJest, pathToConfig } from './helpers'
import { readFileSync } from 'fs'
import { CoverageMapProvider } from './Coverage'
import { updateDiagnostics, resetDiagnostics, failedSuiteCount } from './diagnostics'
import { DebugCodeLensProvider } from './DebugCodeLens'
import { NodeDebugConfigurationProvider, SnippetDebugConfigurationProvider } from './DebugConfigurationProvider'
import { DecorationOptions } from './types'
import { hasDocument, isOpenInMultipleEditors } from './editor'
import { CoverageOverlay } from './Coverage/CoverageOverlay'
import { JestProcess, JestProcessManager } from './JestProcessManagement'

export class JestExt {
  private workspace: ProjectWorkspace
  private jestSettings: Settings
  private pluginSettings: IPluginSettings

  coverageMapProvider: CoverageMapProvider
  coverageOverlay: CoverageOverlay

  testResultProvider: TestResultProvider
  public debugCodeLensProvider: DebugCodeLensProvider
  nodeDebugConfigurationProvider: NodeDebugConfigurationProvider
  snippetDebugConfigurationProvider: SnippetDebugConfigurationProvider

  // So you can read what's going on
  private channel: vscode.OutputChannel

  // The ability to show fails in the problems section
  private failDiagnostics: vscode.DiagnosticCollection

  private passingItStyle: vscode.TextEditorDecorationType
  private failingItStyle: vscode.TextEditorDecorationType
  private skipItStyle: vscode.TextEditorDecorationType
  private unknownItStyle: vscode.TextEditorDecorationType

  private parsingTestFile = false

  // We have to keep track of our inline assert fails to remove later
  failingAssertionDecorators: { [fileName: string]: vscode.TextEditorDecorationType[] }

  private jestProcessManager: JestProcessManager
  private jestProcess: JestProcess

  private clearOnNextInput: boolean

  constructor(workspace: ProjectWorkspace, outputChannel: vscode.OutputChannel, pluginSettings: IPluginSettings) {
    this.workspace = workspace
    this.channel = outputChannel
    this.failingAssertionDecorators = {}
    this.failDiagnostics = vscode.languages.createDiagnosticCollection('Jest')
    this.clearOnNextInput = true
    this.jestSettings = new Settings(workspace)
    this.pluginSettings = pluginSettings

    this.coverageMapProvider = new CoverageMapProvider()
    this.coverageOverlay = new CoverageOverlay(this.coverageMapProvider, pluginSettings.showCoverageOnLoad)

    this.testResultProvider = new TestResultProvider()
    this.debugCodeLensProvider = new DebugCodeLensProvider(this.testResultProvider, pluginSettings.enableCodeLens)
    this.nodeDebugConfigurationProvider = new NodeDebugConfigurationProvider()
    this.snippetDebugConfigurationProvider = new SnippetDebugConfigurationProvider()

    this.jestProcessManager = new JestProcessManager({
      projectWorkspace: workspace,
      runAllTestsFirstInWatchMode: this.pluginSettings.runAllTestsFirst,
    })

    this.getSettings()
    // The theme stuff
    this.setupDecorators()
    // The bottom bar thing
    this.setupStatusBar()
    //reset the jest diagnostics
    resetDiagnostics(this.failDiagnostics)

    // If we should start the process by default, do so
    if (this.pluginSettings.autoEnable) {
      this.startProcess()
    } else {
      this.channel.appendLine('Skipping initial Jest runner process start.')
    }
  }

  private handleStdErr(error: Buffer) {
    const message = error.toString()

    if (this.shouldIgnoreOutput(message)) {
      return
    }

    // The "tests are done" message comes through stdErr
    // We want to use this as a marker that the console should
    // be cleared, as the next input will be from a new test run.
    if (this.clearOnNextInput) {
      this.clearOnNextInput = false
      this.parsingTestFile = false
      this.channel.clear()
    }
    // thanks Qix, http://stackoverflow.com/questions/25245716/remove-all-ansi-colors-styles-from-strings
    const noANSI = message.replace(/[\u001b\u009b][[()#;?]*(?:[0-9]{1,4}(?:;[0-9]{0,4})*)?[0-9A-ORZcf-nqry=><]/g, '')
    if (noANSI.includes('snapshot test failed')) {
      this.detectedSnapshotErrors()
    }

    this.channel.appendLine(noANSI)
  }

  private assignHandlers(jestProcess) {
    jestProcess
      .onJestEditorSupportEvent('executableJSON', (data: JestTotalResults) => {
        this.updateWithData(data)
      })
      .onJestEditorSupportEvent('executableOutput', (output: string) => {
        if (!this.shouldIgnoreOutput(output)) {
          this.channel.appendLine(output)
        }
      })
      .onJestEditorSupportEvent('executableStdErr', (error: Buffer) => this.handleStdErr(error))
      .onJestEditorSupportEvent('nonTerminalError', (error: string) => {
        this.channel.appendLine(`Received an error from Jest Runner: ${error.toString()}`)
      })
      .onJestEditorSupportEvent('exception', result => {
        this.channel.appendLine(`\nException raised: [${result.type}]: ${result.message}\n`)
      })
      .onJestEditorSupportEvent('terminalError', (error: string) => {
        this.channel.appendLine('\nException raised: ' + error)
      })
  }

  public startProcess() {
    if (this.jestProcessManager.numberOfProcesses > 0) {
      return
    }

    if (this.pluginSettings.runAllTestsFirst) {
      this.testsHaveStartedRunning()
    }

    this.jestProcess = this.jestProcessManager.startJestProcess({
      watch: true,
      keepAlive: true,
      exitCallback: (jestProcess, jestProcessInWatchMode) => {
        if (jestProcessInWatchMode) {
          this.jestProcess = jestProcessInWatchMode

          this.channel.appendLine('Finished running all tests. Starting watch mode.')
          status.running('Starting watch mode')

          this.assignHandlers(this.jestProcess)
        } else {
          status.stopped()
          if (!jestProcess.stopRequested) {
            this.channel.appendLine(
              'Starting Jest in Watch mode failed too many times and has been stopped. Please check your system configuration.'
            )
          }
        }
      },
    })

    this.assignHandlers(this.jestProcess)
  }

  public stopProcess() {
    this.channel.appendLine('Closing Jest')
    this.jestProcessManager.stopAll()
    status.stopped()
  }

  private getSettings() {
    this.getJestVersion(jestVersionMajor => {
      if (jestVersionMajor < 20) {
        vscode.window.showErrorMessage(
          'This extension relies on Jest 20+ features, it will continue to work, but some features may not work correctly.'
        )
      }
      this.workspace.localJestMajorVersion = jestVersionMajor
    })

    // Do nothing for the minute, the above ^ can come back once
    // https://github.com/facebook/jest/pull/3592 is deployed
    try {
      this.jestSettings.getConfig(() => {})
    } catch (error) {
      console.log('[vscode-jest] Getting Jest config crashed, likely due to Jest version being below version 20.')
    }
  }

  private detectedSnapshotErrors() {
    if (!this.pluginSettings.enableSnapshotUpdateMessages) {
      return
    }
    vscode.window
      .showInformationMessage('Would you like to update your Snapshots?', { title: 'Replace them' })
      .then(response => {
        // No response == cancel
        if (response) {
          this.jestProcess.runJestWithUpdateForSnapshots(() => {
            if (this.pluginSettings.restartJestOnSnapshotUpdate) {
              this.jestProcessManager.stopJestProcess(this.jestProcess).then(() => {
                this.startProcess()
              })
              vscode.window.showInformationMessage('Updated Snapshots and restarted Jest.')
            } else {
              vscode.window.showInformationMessage('Updated Snapshots. It will show in your next test run.')
            }
          })
        }
      })
  }

  public triggerUpdateDecorations(editor: vscode.TextEditor) {
    this.coverageOverlay.updateVisibleEditors()

    if (!this.canUpdateDecorators(editor)) {
      return
    }

    // OK - lets go
    this.parsingTestFile = true
    this.updateDotDecorators(editor)
    this.parsingTestFile = false
  }

  public triggerUpdateSettings(updatedSettings: IPluginSettings) {
    this.pluginSettings = updatedSettings

    this.workspace.rootPath = updatedSettings.rootPath
    this.workspace.pathToJest = pathToJest(updatedSettings)
    this.workspace.pathToConfig = pathToConfig(updatedSettings)

    this.jestSettings = new Settings(this.workspace)

    this.coverageOverlay.enabled = updatedSettings.showCoverageOnLoad
    this.debugCodeLensProvider.enabled = updatedSettings.enableCodeLens

    this.stopProcess()

    setTimeout(() => {
      this.startProcess()
    }, 500)
  }

  private updateDotDecorators(editor: vscode.TextEditor) {
    const filePath = editor.document.fileName
    const testResults = this.testResultProvider.getSortedResults(filePath)

    // Dots
    const styleMap = [
      { data: testResults.success, decorationType: this.passingItStyle, state: TestReconciliationState.KnownSuccess },
      { data: testResults.fail, decorationType: this.failingItStyle, state: TestReconciliationState.KnownFail },
      { data: testResults.skip, decorationType: this.skipItStyle, state: TestReconciliationState.KnownSkip },
      { data: testResults.unknown, decorationType: this.unknownItStyle, state: TestReconciliationState.Unknown },
    ]

    styleMap.forEach(style => {
      const decorators = this.generateDotsForItBlocks(style.data, style.state)
      editor.setDecorations(style.decorationType, decorators)
    })

    // Debug CodeLens
    this.debugCodeLensProvider.didChange()

    // Inline error messages
    this.resetInlineErrorDecorators(editor)
    if (this.pluginSettings.enableInlineErrorMessages) {
      const fileName = editor.document.fileName
      testResults.fail.forEach(a => {
        const { style, decorator } = this.generateInlineErrorDecorator(fileName, a)
        editor.setDecorations(style, [decorator])
      })
    }
  }

  private resetInlineErrorDecorators(editor: vscode.TextEditor) {
    if (!this.failingAssertionDecorators[editor.document.fileName]) {
      this.failingAssertionDecorators[editor.document.fileName] = []
      return
    }

    if (isOpenInMultipleEditors(editor.document)) {
      return
    }

    this.failingAssertionDecorators[editor.document.fileName].forEach(element => {
      element.dispose()
    })
    this.failingAssertionDecorators[editor.document.fileName] = []
  }

  private generateInlineErrorDecorator(fileName: string, test: TestResult) {
    const errorMessage = test.terseMessage || test.shortMessage
    const decorator = {
      range: new vscode.Range(test.lineNumberOfError, 0, test.lineNumberOfError, 0),
      hoverMessage: errorMessage,
    }

    // We have to make a new style for each unique message, this is
    // why we have to remove off of them beforehand
    const style = decorations.failingAssertionStyle(errorMessage)
    this.failingAssertionDecorators[fileName].push(style)

    return { style, decorator }
  }

  canUpdateDecorators(editor: vscode.TextEditor) {
    const atEmptyScreen = !editor
    if (atEmptyScreen) {
      return false
    }

    const inSettings = !editor.document
    if (inSettings) {
      return false
    }

    if (this.parsingTestFile) {
      return false
    }

    const isATestFile = this.wouldJestRunURI(editor.document.uri)
    return isATestFile
  }

  private wouldJestRunURI(uri: vscode.Uri) {
    const filePath = uri.fsPath

    const globs: string[] = (this.jestSettings.settings as any).testMatch
    if (globs && globs.length) {
      const matchers = globs.map(each => matcher(each, { dot: true }))
      const matched = matchers.some(isMatch => isMatch(filePath))
      return matched
    }

    const root = this.pluginSettings.rootPath
    let relative = path.normalize(path.relative(root, filePath))
    // replace windows path separator with normal slash
    if (path.sep === '\\') {
      relative = relative.replace(/\\/g, '/')
    }
    const testRegex = new RegExp(this.jestSettings.settings.testRegex)
    const matches = relative.match(testRegex)
    return matches && matches.length > 0
  }

  private setupStatusBar() {
    status.initial()
  }

  private setupDecorators() {
    this.passingItStyle = decorations.passingItName()
    this.failingItStyle = decorations.failingItName()
    this.skipItStyle = decorations.skipItName()
    this.unknownItStyle = decorations.notRanItName()
  }

  private shouldIgnoreOutput(text: string): boolean {
    // this fails when snapshots change - to be revised - returning always false for now
    return text.includes('Watch Usage')
  }

  private testsHaveStartedRunning() {
    this.channel.clear()
    status.running('initial full test run')
  }

  private updateWithData(data: JestTotalResults) {
    const normalizedData = resultsWithLowerCaseWindowsDriveLetters(data)
    this.coverageMapProvider.update(normalizedData.coverageMap)

    const statusList = this.testResultProvider.updateTestResults(normalizedData)
    updateDiagnostics(statusList, this.failDiagnostics)

    const failedFileCount = failedSuiteCount(this.failDiagnostics)
    if (failedFileCount <= 0 && normalizedData.success) {
      status.success()
    } else {
      status.failed(` (${failedFileCount} test suite${failedFileCount > 1 ? 's' : ''} failed)`)
    }

    for (const editor of vscode.window.visibleTextEditors) {
      this.triggerUpdateDecorations(editor)
    }
    this.clearOnNextInput = true
  }

  private generateDotsForItBlocks(blocks: TestResult[], state: TestReconciliationState): DecorationOptions[] {
    const nameForState = {
      [TestReconciliationState.KnownSuccess]: 'Passed',
      [TestReconciliationState.KnownFail]: 'Failed',
      [TestReconciliationState.KnownSkip]: 'Skipped',
      [TestReconciliationState.Unknown]: 'Test has not run yet, due to Jest only running tests related to changes.',
    }

    return blocks.map(it => {
      return {
        range: new vscode.Range(it.start.line, it.start.column, it.start.line, it.start.column + 1),
        hoverMessage: nameForState[state],

        /* ERROR: this needs to include all ancestor describe block names as well!
          in code bellow it block has identifier = 'aaa bbb ccc': but name is only 'ccc'

          describe('aaa', () => {
            describe('bbb', () => {
              it('ccc', () => {
              });
            });
          });
        */
        identifier: it.name,
      }
    })
  }

  public deactivate() {
    this.jestProcessManager.stopAll()
  }

  private getJestVersion(version: (v: number) => void) {
    let ver = 18 // default to the last pre-20 release if nothing else can be determined
    const packageJSON = pathToJestPackageJSON(this.pluginSettings)

    if (packageJSON) {
      const contents = readFileSync(packageJSON, 'utf8')
      const packageMetadata = JSON.parse(contents)

      if (packageMetadata['version']) {
        ver = parseInt(packageMetadata['version'])
      }
    }

    version(ver)
  }

  public runTest = async (fileName: string, identifier: string) => {
    const restart = this.jestProcessManager.numberOfProcesses > 0
    this.jestProcessManager.stopAll()

    this.nodeDebugConfigurationProvider.prepareTestRun(fileName, identifier)

    const handle = vscode.debug.onDidTerminateDebugSession(_ => {
      handle.dispose()
      if (restart) {
        this.startProcess()
      }
    })

    const workspaceFolder = vscode.workspace.workspaceFolders[0]
    try {
      // try to run the debug configuration from launch.json
      await vscode.debug.startDebugging(workspaceFolder, 'vscode-jest-tests')
    } catch {
      // if that fails, there (probably) isn't any debug configuration (at least no correctly named one)
      // therefore run the test using the default configuration
      const debugConfiguration = this.snippetDebugConfigurationProvider.provideDebugConfigurations(workspaceFolder)[0]
      await vscode.debug.startDebugging(workspaceFolder, debugConfiguration)
    }
  }

  onDidCloseTextDocument(document: vscode.TextDocument) {
    this.removeCachedTestResults(document)
    this.removeCachedDecorationTypes(document)
  }

  removeCachedTestResults(document: vscode.TextDocument) {
    if (!document || document.isUntitled) {
      return
    }

    const filePath = document.fileName
    this.testResultProvider.removeCachedResults(filePath)
  }

  removeCachedDecorationTypes(document: vscode.TextDocument) {
    if (!document || !document.fileName) {
      return
    }

    delete this.failingAssertionDecorators[document.fileName]
  }

  onDidChangeActiveTextEditor(editor: vscode.TextEditor) {
    if (!hasDocument(editor)) {
      return
    }

    this.triggerUpdateDecorations(editor)
  }

  /**
   * This event is fired with the document not dirty when:
   * - before the onDidSaveTextDocument event
   * - the document was changed by an external editor
   */
  onDidChangeTextDocument(event: vscode.TextDocumentChangeEvent) {
    if (event.document.isDirty) {
      return
    }
    if (event.document.uri.scheme === 'git') {
      return
    }

    // Ignore a clean file with a change:
    if (event.contentChanges.length > 0) {
      return
    }

    this.removeCachedTestResults(event.document)

    for (const editor of vscode.window.visibleTextEditors) {
      if (editor.document === event.document) {
        this.triggerUpdateDecorations(editor)
      }
    }
  }

  toggleCoverageOverlay() {
    this.coverageOverlay.toggleVisibility()
  }
}
