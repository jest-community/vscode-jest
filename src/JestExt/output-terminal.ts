import * as vscode from 'vscode';
import { ExtErrorDef } from '../errors';

/**
 * This class write out Jest run output to vscode.Terminal
 */

export interface JestExtOutput {
  write: (msg: string, opt?: OutputOptions) => string;
}

/**
 * simple class to manage the pending data before the terminal is
 * visible.
 * The default is to keep max 100 output batch (each push counts 1), when exceeding
 * it will remove the first 10 batches. We could make this configurable
 * if needed.
 */
export class PendingOutput {
  private pendingOutput: string[];
  constructor(private maxLength = 100) {
    this.pendingOutput = [];
  }
  push(output: string): void {
    if (this.pendingOutput.length >= this.maxLength) {
      // truncate the first few blocks to make room for the new output
      const cutoff = Math.max(Math.floor(this.maxLength / 10), 1);
      this.pendingOutput = this.pendingOutput.slice(cutoff);
    }
    this.pendingOutput.push(output);
  }
  clear(): void {
    this.pendingOutput = [];
  }
  toString(): string {
    return this.pendingOutput.join('');
  }
}

/** termerinal per workspace */
export class ExtOutputTerminal implements JestExtOutput {
  private pendingMessages: PendingOutput;
  private ptyIsOpen: boolean;
  private writeEmitter = new vscode.EventEmitter<string>();
  private _terminal?: vscode.Terminal;
  private canReveal: boolean;
  public revealOnError: boolean;

  private pty: vscode.Pseudoterminal = {
    onDidWrite: this.writeEmitter.event,
    open: () => {
      this.writeEmitter.fire(`${this.name}\r\n`);
      const pending = this.pendingMessages.toString();
      if (pending) {
        this.writeEmitter.fire(pending);
        this.pendingMessages.clear();
      }
      this.ptyIsOpen = true;
    },
    close: () => {
      this.ptyIsOpen = false;
      this.canReveal = false;
      this._terminal = undefined;
    },
  };
  constructor(private name: string, visibile?: boolean) {
    this.ptyIsOpen = false;
    this.pendingMessages = new PendingOutput();
    this.canReveal = visibile ?? false;
    this.revealOnError = true;
  }

  /**
   * indicate the terminal can be revealed if needed.
   * This allows the terminal to be created in a "background" mode, i.e. it will not force the terminal to be shown if the panels are hidden or
   * if there are other active terminal
   * @returns
   */
  reveal(): void {
    if (this.canReveal) {
      return;
    }
    this.canReveal = true;
    this.createTerminalIfNeeded();
  }

  /** delay creating terminal until we are actually running the tests */
  private createTerminalIfNeeded() {
    if (!this.canReveal || this._terminal) {
      return;
    }
    vscode.window.terminals.forEach((t) => {
      if (t.name === this.name) {
        t.dispose();
      }
    });
    this._terminal = vscode.window.createTerminal({
      name: this.name,
      iconPath: new vscode.ThemeIcon('beaker'),
      isTransient: true,
      pty: this.pty,
    });
    this.ptyIsOpen = false;
  }
  private appendRaw(text: string): void {
    //ensure terminal is created
    this.createTerminalIfNeeded();

    if (this.ptyIsOpen) {
      this.writeEmitter.fire(text);
    } else {
      this.pendingMessages.push(text);
    }
  }

  write(msg: string, opt?: OutputOptions): string {
    const text = toAnsi(msg, opt);
    this.appendRaw(text);

    if (isErrorOutputType(opt) && this.revealOnError) {
      this.show();
    }
    return text;
  }
  show(): void {
    this.reveal();
    this._terminal?.show(true);
  }
  close(): void {
    this._terminal?.dispose();
  }
  dispose(): void {
    this.writeEmitter.dispose();
    this._terminal?.dispose();
  }
}
export class JestOutputTerminal extends ExtOutputTerminal {
  constructor(workspaceName: string, visible?: boolean) {
    super(`Jest (${workspaceName})`, visible);
  }
}

export const AnsiSeq = {
  error: '\x1b[0;31m',
  success: '\x1b[0;32m',
  warn: '\x1b[0;33m',
  info: '\x1b[0;34m',
  bold: '\x1b[1m',
  end: '\x1b[0m',
  lf: '\r\n',
};
export type AnsiSeqType = keyof typeof AnsiSeq;

export const ansiEsc = (type: AnsiSeqType, text: string): string => {
  return `${AnsiSeq[type]}${text}${AnsiSeq.end}`;
};

export type OutputOptionShort =
  | 'error'
  | 'warn'
  | 'new-line'
  | 'bold'
  | 'info'
  | 'success'
  | 'lite';
export type OutputOptions =
  | Array<OutputOptionShort | ExtErrorDef>
  | OutputOptionShort
  | ExtErrorDef;

const isErrorOutputType = (options?: OutputOptions): boolean => {
  if (!options) {
    return false;
  }
  if (Array.isArray(options)) {
    return options.some((opt) => isErrorOutputType(opt) === true);
  }
  if (typeof options === 'string') {
    return options === 'error';
  }
  return options.type === 'error';
};
/** convert string to ansi-coded string based on the options */
const applyAnsiSeq = (text: string, opt: OutputOptionShort, useLite = false): string => {
  switch (opt) {
    case 'error':
    case 'warn':
    case 'info':
    case 'success':
      if (useLite) {
        return `${AnsiSeq[opt]}${text}${AnsiSeq.end}${AnsiSeq.lf}`;
      }
      return `${AnsiSeq.lf}${AnsiSeq[opt]}[${opt}] ${text}${AnsiSeq.end}${AnsiSeq.lf}`;
    case 'bold':
      return `${AnsiSeq[opt]}${text}${AnsiSeq.end}`;
    case 'new-line':
      return `${AnsiSeq.lf}${text}${AnsiSeq.lf}`;
    default:
      return text;
  }
};

export const toAnsi = (msg: string, options?: OutputOptions): string =>
  _toAnsi(msg, options, false);
const _toAnsi = (msg: string, options?: OutputOptions, useLite = false): string => {
  let text = msg.replace(/\n/g, '\r\n');
  if (!options) {
    return text;
  }
  if (Array.isArray(options)) {
    const _useLite = useLite || options.includes('lite');
    return options.reduce((t, opt) => _toAnsi(t, opt, _useLite), msg);
  }

  if (typeof options === 'string') {
    text = applyAnsiSeq(text, options, useLite);
  } else {
    text = applyAnsiSeq(text, options.type, useLite);
    text = `${text}${AnsiSeq.lf}${AnsiSeq.info}[info]${AnsiSeq.end} ${options.desc}, please see: ${options.helpLink}${AnsiSeq.lf}`;
  }
  return text;
};
