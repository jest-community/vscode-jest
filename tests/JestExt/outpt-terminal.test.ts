jest.unmock('../../src/JestExt/output-terminal');
jest.unmock('../../src/errors');

import * as vscode from 'vscode';
import { ansiEsc, AnsiSeq, JestOutputTerminal, toAnsi } from '../../src/JestExt/output-terminal';
import * as errors from '../../src/errors';

describe('JestOutputTerminal', () => {
  let mockTerminal;
  let mockEmitter;

  beforeEach(() => {
    jest.resetAllMocks();

    mockTerminal = {
      dispose: jest.fn(),
      show: jest.fn(),
    };
    vscode.window.createTerminal = jest.fn().mockReturnValue(mockTerminal);
    (vscode.window.terminals as any) = [];

    mockEmitter = { fire: jest.fn(), event: jest.fn(), dispose: jest.fn() };
    (vscode.EventEmitter as jest.Mocked<any>).mockImplementation(() => mockEmitter);
  });
  it('delay creating terminal until the actual write occurs', () => {
    const output = new JestOutputTerminal('workspace');
    expect(vscode.window.createTerminal).not.toBeCalled();
    output.write('abc');
    expect(vscode.window.createTerminal).toBeCalled();
  });
  it('if terminal already opened, close and create a new one', () => {
    const a = { name: 'Jest (a)', dispose: jest.fn() };
    const b = { name: 'Jest (b)', dispose: jest.fn() };
    (vscode.window.terminals as any) = [a, b];
    const t = new JestOutputTerminal('a');
    expect(vscode.window.createTerminal).not.toBeCalled();
    expect(a.dispose).not.toBeCalled();

    t.write('something');
    expect(vscode.window.createTerminal).toBeCalled();
    expect(a.dispose).toBeCalled();
    expect(b.dispose).not.toBeCalled();
  });
  it('can buffer output until open', () => {
    const output = new JestOutputTerminal('a');
    output.write('text 1');
    expect(mockEmitter.fire).not.toBeCalled();

    // after open, the buffered text should be sent again
    const { pty } = (vscode.window.createTerminal as jest.Mocked<any>).mock.calls[0][0];
    pty.open();
    expect(mockEmitter.fire).toBeCalledWith('text 1');

    output.write('text 2');
    expect(mockEmitter.fire).toBeCalledWith('text 2');
  });
  it('if user close the terminal, it will be reopened on the next write', () => {
    const output = new JestOutputTerminal('a');
    output.write('1');
    expect(vscode.window.createTerminal).toBeCalledTimes(1);
    const { pty } = (vscode.window.createTerminal as jest.Mocked<any>).mock.calls[0][0];

    // simulate users close the terminal
    pty.close();
    expect(mockTerminal.dispose).toBeCalled();

    // user writes again
    output.write('1');
    // terminal should be opened again with the same pty
    expect(vscode.window.createTerminal).toBeCalledTimes(2);
    const { pty: pty2 } = (vscode.window.createTerminal as jest.Mocked<any>).mock.calls[1][0];
    expect(pty2).toBe(pty);
  });
  it('will show terminal when writing error messages', () => {
    const output = new JestOutputTerminal('a');

    output.write('1');
    expect(mockTerminal.show).not.toBeCalled();

    output.write('2', 'error');
    expect(mockTerminal.show).toBeCalledTimes(1);

    output.write('2', errors.GENERIC_ERROR);
    expect(mockTerminal.show).toBeCalledTimes(2);
  });
  it('will properly dispose terminal and emitter', () => {
    const output = new JestOutputTerminal('a');
    output.write('1');
    output.dispose();
    expect(mockTerminal.dispose).toBeCalled();
    expect(mockEmitter.dispose).toBeCalled();
  });
  describe('can write output with options', () => {
    it.each`
      case | text                        | options
      ${1} | ${'regular text'}           | ${undefined}
      ${2} | ${'text with newline\r\n'}  | ${undefined}
      ${3} | ${'error text'}             | ${'error'}
      ${4} | ${'warning text'}           | ${'warn'}
      ${5} | ${'bold text'}              | ${'bold'}
      ${6} | ${'bold text with newLine'} | ${['bold', 'new-line']}
    `('can write with option: case $case', ({ text, options }) => {
      const output = new JestOutputTerminal('a');
      const t = output.write(text, options);
      expect(t).toMatchSnapshot();
    });
  });
});
describe('text format utility function', () => {
  it.each`
    options
    ${'error'}
    ${'warn'}
    ${'new-line'}
    ${'bold'}
    ${'info'}
    ${'success'}
    ${'lite'}
    ${['error', 'lite']}
    ${['bold', 'new-line']}
  `('toAnsi: format by output options: $options', ({ options }) => {
    const t = toAnsi('a message', options);
    expect(t).toMatchSnapshot();
  });
  it.each`
    desc         | escSeq
    ${'error'}   | ${AnsiSeq.error}
    ${'success'} | ${AnsiSeq.success}
    ${'warn'}    | ${AnsiSeq.warn}
    ${'info'}    | ${AnsiSeq.info}
    ${'bold'}    | ${AnsiSeq.bold}
    ${'lf'}      | ${AnsiSeq.lf}
  `('ansiEsc: format by ANSI escape sequence: $desc', ({ escSeq }) => {
    expect(ansiEsc(escSeq, 'whatever')).toMatchSnapshot();
  });
});
