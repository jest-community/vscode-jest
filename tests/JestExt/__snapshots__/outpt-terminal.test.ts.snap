// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JestOutputTerminal can write output with options can write with option: case 1 1`] = `"regular text"`;

exports[`JestOutputTerminal can write output with options can write with option: case 2 1`] = `
"text with newline

"
`;

exports[`JestOutputTerminal can write output with options can write with option: case 3 1`] = `
"
[0;31m[error] error text[0m
"
`;

exports[`JestOutputTerminal can write output with options can write with option: case 4 1`] = `
"
[0;33m[warn] warning text[0m
"
`;

exports[`JestOutputTerminal can write output with options can write with option: case 5 1`] = `"[1mbold text[0m"`;

exports[`JestOutputTerminal can write output with options can write with option: case 6 1`] = `
"
[1mbold text with newLine[0m
"
`;

exports[`text format utility function ansiEsc: format by ANSI escape sequence: bold 1`] = `"undefinedwhatever[0m"`;

exports[`text format utility function ansiEsc: format by ANSI escape sequence: error 1`] = `"undefinedwhatever[0m"`;

exports[`text format utility function ansiEsc: format by ANSI escape sequence: info 1`] = `"undefinedwhatever[0m"`;

exports[`text format utility function ansiEsc: format by ANSI escape sequence: lf 1`] = `"undefinedwhatever[0m"`;

exports[`text format utility function ansiEsc: format by ANSI escape sequence: success 1`] = `"undefinedwhatever[0m"`;

exports[`text format utility function ansiEsc: format by ANSI escape sequence: warn 1`] = `"undefinedwhatever[0m"`;

exports[`text format utility function toAnsi: format by output options: ["bold", "new-line"] 1`] = `
"
[1ma message[0m
"
`;

exports[`text format utility function toAnsi: format by output options: ["error", "lite"] 1`] = `
"[0;31ma message[0m

"
`;

exports[`text format utility function toAnsi: format by output options: bold 1`] = `"[1ma message[0m"`;

exports[`text format utility function toAnsi: format by output options: error 1`] = `
"
[0;31m[error] a message[0m
"
`;

exports[`text format utility function toAnsi: format by output options: info 1`] = `
"
[0;34m[info] a message[0m
"
`;

exports[`text format utility function toAnsi: format by output options: lite 1`] = `"a message"`;

exports[`text format utility function toAnsi: format by output options: new-line 1`] = `
"
a message
"
`;

exports[`text format utility function toAnsi: format by output options: success 1`] = `
"
[0;32m[success] a message[0m
"
`;

exports[`text format utility function toAnsi: format by output options: warn 1`] = `
"
[0;33m[warn] a message[0m
"
`;
